pkgs/development/libraries/libxcrypt/default.nix

```
@@ -28,6 +28,16 @@ stdenv.mkDerivation (finalAttrs: {
     "--disable-werror"
   ];
 
+  # lld 17 sets `--no-undefined-version` by default and `libxcrypt`'s version
+  # script unconditionally lists legacy compatibility symbols, even when they're
+  # not exported: https://github.com/besser82/libxcrypt/issues/181
+  env = let
+    lld17Plus = stdenv.cc.bintools.isLLVM
+      && lib.versionAtLeast stdenv.cc.bintools.version "17";
+  in {
+    NIX_LDFLAGS = lib.optionalString lld17Plus "--undefined-version";
+  };
```

nit: using optonalAttrs to env will not cause rebuilds. as would conditionally adding `LDFLAGS` to the configureFlags i think.
```suggestion
  env = let
    lld17Plus = stdenv.cc.bintools.isLLVM
      && lib.versionAtLeast stdenv.cc.bintools.version "17";
  in lib.optionalAttrs lld17Plus {
    NIX_LDFLAGS = "--undefined-version";
  };
```

---

^ is similar to what I started with; unfortunately it yields strange infinite recursion errors:

<details> <summary>Click to expand</summary>

```console
$ nix-instantiate . -A pkgsLLVM.libxcrypt --show-trace
error:
       … while evaluating call site

       at /src/nixpkgs/pkgs/top-level/all-packages.nix:23180:15:

        23179|
        23180|   libxcrypt = callPackage ../development/libraries/libxcrypt {
             |               ^
        23181|     fetchurl = stdenv.fetchurlBoot;

       … while calling 'callPackageWith'

       at /src/nixpkgs/lib/customisation.nix:212:35:

          211|   */
          212|   callPackageWith = autoArgs: fn: args:
             |                                   ^
          213|     let

       … while evaluating call site

       at /src/nixpkgs/lib/customisation.nix:264:13:

          263|     in if missingArgs == {}
          264|        then makeOverridable f allArgs
             |             ^
          265|        # This needs to be an abort so it can't be caught with `builtins.tryEval`,

       … while calling anonymous lambda

       at /src/nixpkgs/lib/customisation.nix:136:17:

          135|     in
          136|     mirrorArgs (origArgs:
             |                 ^
          137|     let

       … while evaluating call site

       at /src/nixpkgs/lib/customisation.nix:138:16:

          137|     let
          138|       result = f origArgs;
             |                ^
          139|

       … while calling anonymous lambda

       at /src/nixpkgs/pkgs/development/libraries/libxcrypt/default.nix:1:1:

            1| { lib, stdenv, fetchurl, perl
             | ^
            2| # Update the enabled crypt scheme ids in passthru when the enabled hashes change

       … while evaluating call site

       at /src/nixpkgs/pkgs/development/libraries/libxcrypt/default.nix:9:1:

            8|
            9| stdenv.mkDerivation (finalAttrs: {
             | ^
           10|   pname = "libxcrypt";

       … while calling 'mkDerivation'

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:46:5:

           45|   mkDerivation =
           46|     fnOrAttrs:
             |     ^
           47|       if builtins.isFunction fnOrAttrs

       … while evaluating call site

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:48:12:

           47|       if builtins.isFunction fnOrAttrs
           48|       then makeDerivationExtensible fnOrAttrs
             |            ^
           49|       else makeDerivationExtensibleConst fnOrAttrs;

       … while calling 'makeDerivationExtensible'

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:59:30:

           58|   # Based off lib.makeExtensible, with modifications:
           59|   makeDerivationExtensible = rattrs:
             |                              ^
           60|     let

       … while evaluating call site

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:92:9:

           91|       finalPackage =
           92|         mkDerivationSimple overrideAttrs args;
             |         ^
           93|

       … while calling 'mkDerivationSimple'

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:508:1:

          507| # * [package attributes]: https://nixos.org/manual/nix/stable/glossary#package-attribute-set
          508| {
             | ^
          509|

       … while evaluating call site

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:585:1:

          584|
          585| extendDerivation
             | ^
          586|   validity.handled

       … while calling 'extendDerivation'

       at /src/nixpkgs/lib/customisation.nix:339:43:

          338|   */
          339|   extendDerivation = condition: passthru: drv:
             |                                           ^
          340|     let

       … while evaluating call site

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:549:12:

          548|         (["meta" "passthru" "pos"]
          549|         ++ optional (__structuredAttrs || envIsExportable) "env"
             |            ^
          550|         )

       … while calling 'optional'

       at /src/nixpkgs/lib/lists.nix:784:20:

          783|   */
          784|   optional = cond: elem: if cond then [elem] else [];
             |                    ^
          785|

       … while evaluating call site

       at /src/nixpkgs/pkgs/development/libraries/libxcrypt/default.nix:37:6:

           36|       && lib.versionAtLeast stdenv.cc.bintools.version "17";
           37|   in lib.optionalAttrs lld17Plus {
             |      ^
           38|     NIX_LDFLAGS = "--undefined-version";

       … while calling 'optionalAttrs'

       at /src/nixpkgs/lib/attrsets.nix:1352:5:

         1351|     cond:
         1352|     as:
             |     ^
         1353|     if cond then as else {};

       … while evaluating the attribute 'cc.bintools.isLLVM'

       at /src/nixpkgs/pkgs/stdenv/generic/default.nix:165:14:

          164|
          165|       inherit cc hasCC;
             |              ^
          166|

       … while evaluating the attribute 'llvmPackages.clang'

       at /src/nixpkgs/pkgs/development/compilers/llvm/17/default.nix:125:5:

          124|     # pick clang appropriate for package set we are targeting
          125|     clang =
             |     ^
          126|       /**/ if stdenv.targetPlatform.useLLVM or false then tools.clangUseLLVM

       … while evaluating the attribute 'clangUseLLVM'

       at /src/nixpkgs/pkgs/development/compilers/llvm/17/default.nix:207:5:

          206|
          207|     clangUseLLVM = wrapCCWith rec {
             |     ^
          208|       cc = tools.clang-unwrapped;

       … while evaluating call site

       at /src/nixpkgs/pkgs/development/compilers/llvm/17/default.nix:207:20:

          206|
          207|     clangUseLLVM = wrapCCWith rec {
             |                    ^
          208|       cc = tools.clang-unwrapped;

       … while calling 'wrapCCWith'

       at /src/nixpkgs/pkgs/top-level/all-packages.nix:17063:5:

        17062|   wrapCCWith =
        17063|     { cc
             |     ^
        17064|     , # This should be the only bintools runtime dep with this sort of logic. The

       … while evaluating call site

       at /src/nixpkgs/pkgs/top-level/all-packages.nix:17082:7:

        17081|     } @ extraArgs:
        17082|       callPackage ../build-support/cc-wrapper (let self = {
             |       ^
        17083|     nativeTools = stdenv.targetPlatform == stdenv.hostPlatform && stdenv.cc.nativeTools or false;

       … while calling 'callPackageWith'

       at /src/nixpkgs/lib/customisation.nix:212:35:

          211|   */
          212|   callPackageWith = autoArgs: fn: args:
             |                                   ^
          213|     let

       … while evaluating call site

       at /src/nixpkgs/lib/customisation.nix:264:13:

          263|     in if missingArgs == {}
          264|        then makeOverridable f allArgs
             |             ^
          265|        # This needs to be an abort so it can't be caught with `builtins.tryEval`,

       … while calling anonymous lambda

       at /src/nixpkgs/lib/customisation.nix:136:17:

          135|     in
          136|     mirrorArgs (origArgs:
             |                 ^
          137|     let

       … while evaluating call site

       at /src/nixpkgs/lib/customisation.nix:138:16:

          137|     let
          138|       result = f origArgs;
             |                ^
          139|

       … while calling anonymous lambda

       at /src/nixpkgs/pkgs/build-support/cc-wrapper/default.nix:8:1:

            7|
            8| { name ? ""
             | ^
            9| , lib

       … while evaluating the attribute 'buildInputs' of the derivation 'glibc-x86_64-unknown-linux-gnu-2.39-5'

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:331:7:

          330|     // (optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          331|       name =
             |       ^
          332|         let

       … while evaluating the attribute 'nativeBuildInputs' of the derivation 'linux-headers-6.7'

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:331:7:

          330|     // (optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          331|       name =
             |       ^
          332|         let

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling anonymous lambda

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:302:13:

          301|       (map (drv: getDev drv.__spliced.buildBuild or drv) (checkDependencyList "depsBuildBuild" depsBuildBuild))
          302|       (map (drv: getDev drv.__spliced.buildHost or drv) (checkDependencyList "nativeBuildInputs" nativeBuildInputs'))
             |             ^
          303|       (map (drv: getDev drv.__spliced.buildTarget or drv) (checkDependencyList "depsBuildTarget" depsBuildTarget))

       … while evaluating call site

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:302:18:

          301|       (map (drv: getDev drv.__spliced.buildBuild or drv) (checkDependencyList "depsBuildBuild" depsBuildBuild))
          302|       (map (drv: getDev drv.__spliced.buildHost or drv) (checkDependencyList "nativeBuildInputs" nativeBuildInputs'))
             |                  ^
          303|       (map (drv: getDev drv.__spliced.buildTarget or drv) (checkDependencyList "depsBuildTarget" depsBuildTarget))

       … while calling 'getOutput'

       at /src/nixpkgs/lib/attrsets.nix:1796:23:

         1795|   */
         1796|   getOutput = output: pkg:
             |                       ^
         1797|     if ! pkg ? outputSpecified || ! pkg.outputSpecified

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling anonymous lambda

       at /src/nixpkgs/lib/lists.nix:334:29:

          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                             ^
          335|

       … while evaluating call site

       at /src/nixpkgs/lib/lists.nix:334:32:

          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          335|

       … while calling anonymous lambda

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:276:15:

          275|     imap1
          276|       (index: dep:
             |               ^
          277|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep

       … while evaluating call site

       at /src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:277:12:

          276|       (index: dep:
          277|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |            ^
          278|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … while calling 'isDerivation'

       at /src/nixpkgs/lib/attrsets.nix:1283:5:

         1282|   isDerivation =
         1283|     value: value.type or null == "derivation";
             |     ^
         1284|

       … while evaluating call site

       at /src/nixpkgs/pkgs/top-level/splice.nix:73:14:

           72|             # on to splice them together.
           73|           if lib.isDerivation defaultValue then augmentedValue // spliceReal {
             |              ^
           74|             pkgsBuildBuild = tryGetOutputs valueBuildBuild;

       … while calling 'isDerivation'

       at /src/nixpkgs/lib/attrsets.nix:1283:5:

         1282|   isDerivation =
         1283|     value: value.type or null == "derivation";
             |     ^
         1284|

       … while evaluating the attribute 'perl'

       at /src/nixpkgs/pkgs/top-level/all-packages.nix:25474:3:

        25473|
        25474|   perl = perl538;
             |   ^
        25475|   perlPackages = perl538Packages;

       … while evaluating the attribute 'perl538'

       at /src/nixpkgs/pkgs/development/interpreters/perl/default.nix:66:3:

           65|   # Maint version
           66|   perl538 = callPackage ./interpreter.nix {
             |   ^
           67|     self = perl538;

       … while evaluating call site

       at /src/nixpkgs/pkgs/development/interpreters/perl/default.nix:66:13:

           65|   # Maint version
           66|   perl538 = callPackage ./interpreter.nix {
             |             ^
           67|     self = perl538;

       … while calling 'callPackageWith'

       at /src/nixpkgs/lib/customisation.nix:212:35:

          211|   */
          212|   callPackageWith = autoArgs: fn: args:
             |                                   ^
          213|     let

       … while evaluating call site

       at /src/nixpkgs/lib/customisation.nix:264:13:

          263|     in if missingArgs == {}
          264|        then makeOverridable f allArgs
             |             ^
          265|        # This needs to be an abort so it can't be caught with `builtins.tryEval`,

       … while calling anonymous lambda

       at /src/nixpkgs/lib/customisation.nix:136:17:

          135|     in
          136|     mirrorArgs (origArgs:
             |                 ^
          137|     let

       … while evaluating call site

       at /src/nixpkgs/lib/customisation.nix:138:16:

          137|     let
          138|       result = f origArgs;
             |                ^
          139|

       … while calling anonymous lambda

       at /src/nixpkgs/pkgs/development/interpreters/perl/interpreter.nix:1:1:

            1| { stdenv
             | ^
            2| , fetchurl

       … while evaluating call site

       at /src/nixpkgs/pkgs/top-level/splice.nix:73:14:

           72|             # on to splice them together.
           73|           if lib.isDerivation defaultValue then augmentedValue // spliceReal {
             |              ^
           74|             pkgsBuildBuild = tryGetOutputs valueBuildBuild;

       … while calling 'isDerivation'

       at /src/nixpkgs/lib/attrsets.nix:1283:5:

         1282|   isDerivation =
         1283|     value: value.type or null == "derivation";
             |     ^
         1284|

       … while evaluating the attribute 'libxcrypt'

       at /src/nixpkgs/pkgs/top-level/all-packages.nix:23180:3:

        23179|
        23180|   libxcrypt = callPackage ../development/libraries/libxcrypt {
             |   ^
        23181|     fetchurl = stdenv.fetchurlBoot;

       error: infinite recursion encountered

       at /src/nixpkgs/pkgs/top-level/all-packages.nix:23180:3:

        23179|
        23180|   libxcrypt = callPackage ../development/libraries/libxcrypt {
             |   ^
        23181|     fetchurl = stdenv.fetchurlBoot;
```

</details>

IIUC the path is: `libxcrypt` (host) -> `stdenv.cc` (from `pkgsLLVM`) -> ... -> `glibc` -> `linux-headers-6.7` -> ... -> `perl` (build) -> `libxcrypt` (??)

I'm not sure exactly what's going on here but it seems like any usage of `stdenv.cc.*` needed to evaluate the keys of `libxcrypt`'s derivation yield this cycle error.

Adding it to one of the other flags attrs (did you mean `makeFlags`?) is a good suggestion though; thanks. I'll try that now.

---

> Adding it to one of the other flags attrs (did you mean `makeFlags`?) is a good suggestion though; thanks. I'll try that now.

configure usually will usually incorporate `CFLAGS`, `CXXFLAGS`, and i think `LDFLAGS` from the environment or command line, i believe.

---

Oh interesting! I didn't realize `configure` lets you pass those vars on the command line. TIL.

I ended up using `makeFlags` (I'm not sure how the existing conditional `LDFLAGS` there would interact with `LDFLAGS` passed to `configure`) but I can try switching to `configureFlags` if you feel that'd be more appropriate; PTAL